docs/refs/crd-upgrade-safety.md

```diff
@@ -0,0 +1,207 @@
+# Custom Resource Definition Upgrade Safety
+
+OLM implements several validation checks when upgrading Custom Resource Definitions
+in order to maintain backwards compatibility with previous versions of that CRD.
+These validation checks are known collectively as the CRD Upgrade Safety preflight check.
+
+The CRD Upgrade Safety preflight check consists of several individual field validations
+that are performed before an upgrade to a CRD is allowed to progress on a cluster.
+
+## Disallowed CRD Upgrade Changes
+
+The following changes to an existing CRD will be caught by the CRD Upgrade
+Safety preflight check and prevent the upgrade:
+- The scope changes from Cluster to Namespace or from Namespace to Cluster
+- An existing stored version of the CRD is removed
+- A new required field is added to an existing version of the CRD
+- An existing field is removed from an existing version of the CRD
+- An existing field type is changed in an existing version of the CRD
+- A new default value is added to a field that did not previously have a default value
+- The default value of a field is changed
+- An existing default value of a field is removed
+- New enum restrictions are added to an existing field which did not previously have enum restrictions
+- Existing enum values from an existing field are removed
+- The minimum value of an existing field is incrased in an existing version
+- The maximum value of an existing field is decreased in an existing version
+- Minimum or maximum field constraints are added to a field that did not previously have constraints
+
+If the CRD Upgrade Safety preflight check encounters one of the disallowed upgrade changes,
+it will log an error for each disallowed change detected in the CRD upgrade.
+
+In cases where a change to the CRD does not fall into one of the disallowed change categories
+but is also unable to be properly detected as allowed, the CRD Upgrade Safety preflight check
+will prevent the upgrade and log an error for an "unknown change."
+
+**Note:** the rules for changes to minimum and maximum values apply to `minimum`, `minLength`,
+`minProperties`, `minItems`, `maximum`, `maxLength`, `maxProperties`, and `maxItems` constrants.
+
+## Allowed CRD Upgrade Changes
+
+The following changes to an existing CRD are safe for backwards compatibility and will
+not cause the CRD Upgrade Safety preflight check to halt the upgrade:
+- Adding new enum values to the list of allowed enum values in a field
+- An existing required field is changed to optional in an existing version
+- The minimum value of an existing field is decreased in an existing version
+- The maximum value of an existing field is increased in an existing version
+- A new version of the CRD is added with no modifications to existing versions
```

Comment:
No change needed in this PR, but something that I think we may have overlooked when originally developing the preflight checks is that new versions of the CRD must be backwards compatible with the served versions _OR_ provide conversion logic to prevent data loss and invalidation of resources stored at an older version.

@joelanford I think we may need to revisit the CRD Upgrade Safety preflight checks to ensure we capture behavior to check for breaking changes between all stored and served versions of the CRD when there is no conversion strategy specified.

Reply:
given the long discussion this birthed in Slack, do we want to hold off on these docs until that work is done or would we just update this once that's complete? Not sure on what kind of timeline you were imagining all this.

Reply:
We can let the doc through and update it as necessary